### PR TITLE
Dialectic search, support and example

### PR DESCRIPTION
**Pull request by randomir**

### CODE REVIEW

File: hybrid/flow.py [hunk 1]
```diff
@@ -587,13 +588,107 @@ def next(self, states):
         return states[min_idx]
 
 
-class Loop(Runnable):
+class TrackMin(Runnable, traits.SISO):
+    """Tracks (and keeps record of) the "best" (or minimal) :class:`State` seen,
+    according to a metric defined with a `key` function.
```
Comment (JoelPasvolsky) [inline hybrid/flow.py]: If the metric is defined by a key, is the "best" always (or minimal)?

Tracks and records the best :class:`State` according to a metric defined with a `key` function; typically this is the minimal :class:`State`

File: hybrid/flow.py [hunk 2]
```diff
@@ -587,13 +588,107 @@ def next(self, states):
...
+            Best state is judged according to a metric defined with a `key`.
+            `key` can be a `callable` with a signature::
+
+                key :: (State s, Ord k) => s -> k
+
+            or a string holding a key name/path to be extracted from the input
+            state with `operator.attrgetter` method.
+
+            By default, `key == operator.attrgetter('samples.first.energy')`,
+            thus favoring states containing a sample with the minimal energy.
```
Comment (JoelPasvolsky) [inline hybrid/flow.py]: Above you use :class:`State` but here you switched to state, and below

File: hybrid/flow.py [hunk 3]
```diff
@@ -587,13 +588,107 @@ def next(self, states):
...
+
+        output (bool, optional, default=False):
+            Update the output state with the relevant aspect of the best state
```
Comment (JoelPasvolsky) [inline hybrid/flow.py]: "the relevant aspect" --> unclear

File: hybrid/flow.py [hunk 4]
```diff
@@ -587,13 +588,107 @@ def next(self, states):
...
+            seen so far.
+
+        input_key (str, optional, default='samples')
+            If `output=True`, then this defines the variable/key name in the
+            input State that shall be included in the output state.
+
+        output_key (str, optional, default='best_samples')
+            If `output=True`, then the key under which the `input_key` from the
+            best state seen so far is stored in the output State.
+
```
Comment (JoelPasvolsky) [inline hybrid/flow.py]: Here you have both "state" and "State" --> this is getting confusing

File: hybrid/flow.py [hunk 5]
```diff
@@ -587,13 +588,107 @@ def next(self, states):
...
-    `convergence` time."""
+    `convergence` number of iterations.
+
+    Args:
+        runnable (:class:`Runnable`):
+            A runnable that's looped over.
+
+        max_iter (int/None, optional, default=1000):
+            Maximum number of times the `runnable` is run, regardless of other
+            termination criteria. This is the upper bound. Can be set to `None`.
```
Comment (JoelPasvolsky) [inline hybrid/flow.py]: "Can be set to `None`" --> Maybe something like, "If set to None, this parameter is ignored" to state the behavior if set to None?

File: hybrid/flow.py [hunk 6]
```diff
@@ -587,13 +588,107 @@ def next(self, states):
...
+
+        convergence (int, optional, default=10):
+            Terminating condition based on no-output-change count. If the output
```
Comment (JoelPasvolsky) [inline hybrid/flow.py]: Terminates upon reaching this number of iterations with unchanged output.

File: hybrid/flow.py [hunk 7]
```diff
@@ -623,34 +718,128 @@ def __repr__(self):
     def __iter__(self):
         return iter((self.runnable,))
 
+    def iteration_update(self, iterno, cnt, input_state, output_state):
+        """Implement "converge on non-changing output" behavior:
```
Comment (JoelPasvolsky) [inline hybrid/flow.py]: unchanging?

File: hybrid/flow.py [hunk 8]
```diff
@@ -623,34 +718,128 @@ def __repr__(self):
...
+
+          - loop `max_iter` times, but bail-out earlier if output doesn't change
```
Comment (JoelPasvolsky) [inline hybrid/flow.py]: I love the bail out --> will try to reuse this in other places myself

File: hybrid/flow.py [hunk 9]
```diff
@@ -623,34 +718,128 @@ def __repr__(self):
...
+    iterations.
+
+    Args:
+        runnable (:class:`Runnable`):
+            A runnable that's looped over.
+
+        max_tries (int, optional, default=10):
+            Maximum number of times the `runnable` is run for the **same** input
+            state. On each improvement, the better state is used for the next
+            input state, and tries counter is reset.
```
Comment (JoelPasvolsky) [inline hybrid/flow.py]: state, and `max_tries` counter is reset?

File: hybrid/flow.py [hunk 10]
```diff
@@ -623,34 +718,128 @@ def __repr__(self):
...
+            thus favoring states containing a sample with the minimal energy.
+
+    """
+
+    def __init__(self, runnable, max_tries=10, key=None):
+        super(LoopWhileNoImprovement, self).__init__(
+            runnable=runnable, max_iter=None, convergence=max_tries, key=key)
+
+    def iteration_update(self, iterno, cnt, input_state, output_state):
+        """Implement "no-improvement count-down" behavior:
```
Comment (JoelPasvolsky) [inline hybrid/flow.py]: Why not just use the two bullets as sentences without this introduction statement?

File: examples/dialectic-search.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+# limitations under the License.
+
+"""Dialectic search [1][2] algorithm outline: For a given assignment (the thesis), it
+greedily improves it. Then it tries to improve the solution further by
+generating randomized modifications (an antithesis) of the current assignment,
+greedily improving it, and then combining the two assignments to form a new
+assignment, which is also greedily improved (the synthesis). If this new
+assignment is at least as good, it is considered the new current assignment. If
+this process does not result in improvements for a while, then the search moves
+to the modified assignment and continues searching from there.
```
[the remark this answers is not in the document]
Comment (randomir) [reply examples/dialectic-search.py]: Funny thing is I copied that part from the paper that I cited. :smile:

File: hybrid/flow.py (hunk 2 above)
Comment (randomir) [reply hybrid/flow.py]: We don't have a clear standard on this (that I know of) - topic for a discussion. I'll update "state" to "State".

File: hybrid/flow.py (hunk 3 above)
Comment (randomir) [reply hybrid/flow.py]: Fixed.

File: hybrid/flow.py (hunk 4 above)
Comment (randomir) [reply hybrid/flow.py]: Fixed.

File: hybrid/flow.py (hunk 9 above)
Comment (randomir) [reply hybrid/flow.py]: No, "try counter", or "trial counter". A counter that counts tries, up to `max_tries`.

File: hybrid/flow.py (hunk 10 above)
Comment (randomir) [reply hybrid/flow.py]: Adds a bit of structure. I can't start a paragraph with a bullet point. :)

Comment (randomir) [reply hybrid/flow.py]: @JoelPasvolsky Thanks for all the suggestions and improvements!

**Pull request merged**